### PR TITLE
Add initial check support to external modules

### PR DESCRIPTION
**Pull request by acammack-r7**
~**Requires #10002 for refactor goodness**~

This adds the infrastructure needed for external modules to report check results, starting with the one exploit template we have. Currently we support `soft_check`s, which are meant to be non-invasive. Framework has historically not differentiated between these and checks which do exploit the target system (`hard_check`s under this naming scheme), but the people I have talked to find this sort of distinction important.

I would like to unify the model around scanners and exploit checks and how that interacts with vuln reporting, but that seems like a task for the future.

Verification
========

- [x] Set up Haraka (https://github.com/rapid7/metasploit-framework/pull/8178)
- [x] `msfconsole`
- [x] `use exploit/linux/smtp/haraka`
- [x] `check IP`
- [x] `[*] REDACTED:25 The target appears to be vulnerable.`
- [x] Check other types of SMTP servers
- [x] They should read as `detected`, since the banner check will fail, but SMTP is up

### CODE REVIEW

File: modules/exploits/linux/smtp/haraka.py [hunk 1]
```diff
@@ -100,8 +106,31 @@ def create_zip(cmd="touch /tmp/harakiri"):
...
+def check_banner(args):
+    module.log('{}:{} Starting banner check for Haraka < 2.8.9'.format(args['rhost'], args['rport']), level='debug')
+    c = smtplib.SMTP()
+    (code, banner) = c.connect(args['rhost'], int(args['rport']))
+    c.quit()
+
+    if code == 220 and 'Haraka' in banner:
+        versions = re.findall('(\d+\.\d+\.\d+)', banner)
+        if versions:
+            if StrictVersion(versions[0]) < StrictVersion('2.8.9'):
```
Comment (bcook-r7) [inline modules/exploits/linux/smtp/haraka.py]: I just like how more formal version checking looks than in python, compared to Ruby where you'll usually find the non-intuitive Gem::Version being used.

Comment (bcook-r7) [reply modules/exploits/linux/smtp/haraka.py]: Maybe it's just the ability to elide namespace with `from`

Comment (acammack-r7) [reply modules/exploits/linux/smtp/haraka.py]: Eh, I don't think this any better/worse. You still have to google to find it, and it's in an equivalent place.

**Comment by bcook-r7** [comment]
## Release Notes

This adds support for external exploit modules to run a check if the target is vulnerable.

**Pull request merged**